include/deal.II/matrix_free/cuda_tensor_product_kernels.h

```diff
@@ -44,6 +44,204 @@ namespace CUDAWrappers
       evaluate_evenodd
     };
 
+
+
+#if KOKKOS_VERSION >= 40000
+    /**
+     * Helper function for values() and gradients() in 1D
+     */
+    template <int n_q_points_1d,
+              typename Number,
+              int  direction,
+              bool dof_to_quad,
+              bool add,
+              bool in_place,
+              typename ViewTypeIn,
+              typename ViewTypeOut>
+    DEAL_II_HOST_DEVICE void
+    apply_1d(const Kokkos::TeamPolicy<
+               MemorySpace::Default::kokkos_space::execution_space>::member_type
+               &team_member,
+             const Kokkos::View<Number *, MemorySpace::Default::kokkos_space>
+                              shape_data,
+             const ViewTypeIn in,
+             ViewTypeOut      out)
+    {
+      Number t[n_q_points_1d];
+      Kokkos::parallel_for(Kokkos::TeamThreadRange(team_member, n_q_points_1d),
+                           [&](const int &q) {
+                             t[q] = 0;
+                             // This loop simply multiplies the shape function
+                             // at the quadrature point by the value finite
+                             // element coefficient.
+                             // FIXME check why using parallel_reduce
+                             // ThreadVector is slower
```

Comment:
What vector lengths have you tried? It only really makes sense to use `ThreadVectorRange` if the vector length is close to the warp/wavefront/subgroup size. Note that the default `vector_length` is `1` which means that you would pretty much be doing the same as now (plus some boilerplate for the reduction algorithm). Of course, all that shouldn't really matter on the CPU.

If the compiler doesn't vectorize the code here already, it might make sense to try to use the `Kokkos::Experimental::simd` types instead.

Reply:
I just used the default value but the results on the GPU were really terrible. I was wondering if there was something wrong when using `TeamThreadMDRange` and `ThreadVector` but I didn't look into the issue that's why there is the FIXME